ios/engine/KMEI/KeymanEngine/Classes/KeymanTextField.swift

```diff
@@ -0,0 +1,355 @@
+      let textRange = selectedTextRange ?? UITextRange()
+      let newRange = NSRange(location: offset(from: beginningOfDocument, to: textRange.start),
+                             length: offset(from: textRange.start, to: textRange.end))
+      Manager.shared.setSelectionRange(newRange, manually: false)
+    }
+  }
+
+  // MARK: - KMWebViewDelegate
+  @objc public func updatedFragment(_ fragment: String) {
+    if fragment.contains("insertText") {
+      processInsertText(fragment)
+    } else if fragment.contains("hideKeyboard") {
+      dismissKeyboard()
+    } else if fragment.contains("menuKeyUp") {
+      if let viewController = viewController {
+        Manager.shared.showKeyboardPicker(in: viewController, shouldAddKeyboard: false)
+      } else {
+        Manager.shared.switchToNextKeyboard()
+      }
+    }
+  }
+
+  private func processInsertText(_ fragment: String) {
+    if Manager.shared.isSubKeysMenuVisible {
+      return
+    }
+
+    if isInputClickSoundEnabled {
+      AudioServicesPlaySystemSound(0x450)
+
+      // Disable input click sound for 0.1 second to ensure it plays for single key stroke.
+      isInputClickSoundEnabled = false
+      perform(#selector(self.enableInputClickSound), with: nil, afterDelay: 0.1)
+    }
+
+    // TODO: Refactor duplicate logic in KeymanInputViewController and KMTextView
+    let dnRange = fragment.range(of: "+dn=")!
+    let sRange = fragment.range(of: "+s=")!
+
+    let dn = Int(fragment[dnRange.upperBound..<sRange.lowerBound])!
+    let s = fragment[sRange.upperBound...]
+
+    let hexStrings: [String]
+    if !s.isEmpty {
+      hexStrings = s.components(separatedBy: ",")
+    } else {
+      hexStrings = []
+    }
+
+    let codeUnits = hexStrings.map { hexString -> UInt16 in
+      let scanner = Scanner(string: hexString)
+      var codeUnit: UInt32 = 0
+      scanner.scanHexInt32(&codeUnit)
+      return UInt16(codeUnit)
+    }
+
+    let text = String(utf16CodeUnits: codeUnits, count: codeUnits.count)
+
+    let textRange = selectedTextRange!
+    let selRange = NSRange(location: offset(from: beginningOfDocument, to: textRange.start),
+                           length: offset(from: textRange.start, to: textRange.end))
+
+    if dn <= 0 {
+      if selRange.length == 0 {
+        insertText(text)
+      } else {
+        self.text = (self.text! as NSString).replacingCharacters(in: selRange, with: text)
+      }
+    } else {
+      if s.isEmpty {
+        for _ in 0..<dn {
+          deleteBackward()
+        }
+      } else {
+        if selRange.length == 0 {
+          for _ in 0..<dn {
+            deleteBackward()
+          }
+          insertText(text)
+        } else {
+          self.text = (self.text! as NSString).replacingCharacters(in: selRange, with: text)
+        }
+      }
+    }
+  }
+
+  // MARK: - UITextFieldDelegate Hooks
+  public override var selectedTextRange: UITextRange? {
+    didSet(range) {
+      guard let range = range else {
+        return
+      }
```

Comment:
There used to be a call to `super` here in Obj-C - is this auto-handled for us somehow?  Is it now unnecessary?

Reply:
We're using `didSet`, which executes after the property is set. `set` isn't overrided, so we don't need to call `super`.